external/java/src/test/java/ch/epfl/dedis/lib/crypto/bn256/BNTest.java

```diff
@@ -0,0 +1,238 @@
+
+            BN.GT minusE2 = new BN.GT().neg(e2);
+            e1.add(e1, minusE2);
+
+            assertFalse(!e1.p.isOne(), "bad pairing result: " + e1.toString());
+        }
+    }
+
+    @Test
+    void g1Marshal() {
+        BN.G1 g = new BN.G1().scalarBaseMul(BigInteger.ONE);
+        byte[] from = g.marshal();
+        assertNotNull(new BN.G1().unmarshal(from));
+
+        g.scalarBaseMul(Constants.order);
+        from = g.marshal();
+        BN.G1 g2 = new BN.G1().unmarshal(from);
+        assertNotNull(g2);
+        assertFalse(!g2.p.isInfinity(), "inf marshaled incorrectly");
+    }
+
+    @Test
+    void g2Marshal() {
+        BN.G2 g = new BN.G2().scalarBaseMul(BigInteger.ONE);
+        byte[] from = g.marshal();
+        assertNotNull(new BN.G2().unmarshal(from));
+
+        g.scalarBaseMul(Constants.order);
+        from = g.marshal();
+        BN.G2 g2 = new BN.G2().unmarshal(from);
+        assertNotNull(g2);
+        assertFalse(!g2.p.isInfinity(), "inf marshaled incorrectly");
+    }
+
+    @Test
+    void g1Identity() {
+        BN.G1 g = new BN.G1().scalarBaseMul(BigInteger.ZERO);
+        assertFalse(!g.p.isInfinity(), "failure");
+    }
+
+    @Test
+    void g2Identity() {
+        BN.G2 g = new BN.G2().scalarBaseMul(BigInteger.ZERO);
+        assertFalse(!g.p.isInfinity(), "failure");
+    }
+
+    @Test
+    void tripartiteDiffieHellman() {
+        Random rnd = new SecureRandom();
+        BigInteger a = BN.randPosBigInt(rnd, Constants.p);
+        BigInteger b = BN.randPosBigInt(rnd, Constants.p);
+        BigInteger c = BN.randPosBigInt(rnd, Constants.p);
+
+        BN.G1 pa = new BN.G1().unmarshal(new BN.G1().scalarBaseMul(a).marshal());
+        BN.G2 qa = new BN.G2().unmarshal(new BN.G2().scalarBaseMul(a).marshal());
+        BN.G1 pb = new BN.G1().unmarshal(new BN.G1().scalarBaseMul(b).marshal());
+        BN.G2 qb = new BN.G2().unmarshal(new BN.G2().scalarBaseMul(b).marshal());
+        BN.G1 pc = new BN.G1().unmarshal(new BN.G1().scalarBaseMul(c).marshal());
+        BN.G2 qc = new BN.G2().unmarshal(new BN.G2().scalarBaseMul(c).marshal());
+
+        BN.GT k1 = BN.pair(pb, qc);
+        k1.scalarMul(k1, a);
+        byte[] k1Bytes = k1.marshal();
+        assertTrue(Arrays.equals(new BN.GT().unmarshal(k1Bytes).marshal(), k1Bytes), "failed to unmarshal GT k1");
+
+        BN.GT k2 = BN.pair(pc, qa);
+        k2.scalarMul(k2, b);
+        byte[] k2Bytes = k2.marshal();
+        assertTrue(Arrays.equals(new BN.GT().unmarshal(k2Bytes).marshal(), k2Bytes), "failed to unmarshal GT k2");
+
+        BN.GT k3 = BN.pair(pa, qb);
+        k3.scalarMul(k3, c);
+        byte[] k3Bytes = k3.marshal();
+        assertTrue(Arrays.equals(new BN.GT().unmarshal(k3Bytes).marshal(), k3Bytes), "failed to unmarshal GT k3");
+
+        assertFalse(!Arrays.equals(k1Bytes, k2Bytes) || !Arrays.equals(k2Bytes, k3Bytes), "keys didn't agree");
+    }
+
+    @Test
+    void benchmarkPairing() {
+        for (int i = 0; i < 10; i++) {
```

Comment:
For a benchmark, that's short... Perhaps should also benchmark multiplications and additions.

Reply:
This is ported from the golang benchmark test. I don't think junit lets you write benchmark tests the same way as go, so my idea was to simply let the use change the value "10" to something else when they actually need to do a benchmark. I can write a comment to explain.